best_practices/api_call_sequencing.md

```diff
@@ -0,0 +1,51 @@
+---
+permalink: /api-call-sequencing
+title: App Installation Design and API Call Sequencing
+layout: default
+section_title: Best Practices
+
+---
+
+## Background ##
+
+As you develop your project, it is important to understand that your design choices and the order in which API calls are made can have a significant impact on the behavior and performance of your application.
+Proper installation design and API call sequencing can help prevent unintended effects when a user with access to more than one company authorizes your application through OAuth 2.0.
+
+## Installation with OAuth 2.0 Authorization
+
+An application not installed from the App Marketplace by a Procore Admin may still be accessed and used by a Procore user if they authorize the application to access their Procore data through the OAuth 2.0 flow.
+The ability to authorize and use an application in a Procore account is controlled by the [Allow User Installs](https://support.procore.com/products/online/user-guide/company-level/admin/tutorials/allow-user-installs) feature in the Procore Web application.
+When a user authorizes the application and subsequently accesses Procore API resources in this manner, it is connected to and visible in the company level App Management tool in Procore.
+If Procore users are able to authorize your application through OAuth 2.0 outside the context of a Marketplace installation, it is important to account for this scenario in your design.
+
+## Handling Multiple Companies
+
+When a user with access to more than one Procore company authorizes your application to access their data through OAuth 2.0, it is important to account for the following scenario in your API call sequencing.
+Consider these conditions:
+
+* Users have access to more than one company in Procore.
+* User authorizes an application to access data on their behalf through OAuth 2.0.
+* Application’s first API call with the user’s access token is to the [List Companies](https://developers.procore.com/reference/rest/v1/companies#list-companies) endpoint (or possibly the second call, after [Show User Info](https://developers.procore.com/reference/rest/v1/me)).
+* Application iterates through the List Companies response body, making subsequent calls to a resource within each company, e.g. [List Projects](https://developers.procore.com/reference/rest/v1/projects#list-projects) (passing in a company_id).
+
+In the above scenario, your application would show up as ‘installed’ in each company you make an API call to (i.e. a request made using a given company’s ID or the ID of a project belonging to that company).
+To prevent this from occurring, we recommend prompting the user to choose the Company they would like to authorize the application to access, and then make the subsequent call to a project resource for only that company.
```

Comment:
@jimj-procoretech Is this the Best Practice? 

How can the developer prompt the user to choose the Company they would like to authorize?
There's no company selector here:

<img width="524" alt="Snap 2021-06-16 at 2 30 37 PM" src="https://user-images.githubusercontent.com/74197530/122437176-50819380-cf4e-11eb-98a8-2472bf6c45cb.png">

Reply:
They would display a modal after the user clicked 'Allow' I presume.

Reply:
As far as I know, there's no modal after the user clicks "Allow".

The user is redirected to the app that has requested the authorization.

Reply:
It would be that callback URL that would put up the modal.

Reply:
@jimj-procoretech I'm still a little confused.

If I build an app and show the modal above the user, then once the user clicks on "Allow" I have obtained the token I need to use Procore on behalf of that user.

Therefore, with the callback URL, I will redirect the user back to my app, where I will be able to display the data fetched from Procore on behalf of that user.

As far as I know, there is no way to **limit** the token obtained from the user to only work for a specific company, in other words I don't think this is feasible in Procore:

> we recommend prompting the user to choose the Company they would like to authorize the application to access,

Reply:
This is the recommendation coming from Ecosystem Product, Platform Product Marketing, and App Validation so my assumption is that this approach was vetted technically and is feasible. It is my understanding that the Marketplace app does this in situations where a company admin (in more than one company) clicks on the 'Login to Install' button on a Marketplace app tile. They are then prompted to select which company they want to access.
Can we please approve this PR in its current form so I can bring closure to this issue and report back to the stakeholders that the document they requested and collaborated on has been published? Thank you.